encrypted-media-respec.html

```diff
@@ -605,8 +605,112 @@
 
           <section>
             <h5>Get Supported Configuration</h5>
-            <p>Given a <a def-id="keysystems"></a> implementation <var>implementation</var>, <a>MediaKeySystemConfiguration</a> <var>candidate configuration</var>, and <var>origin</var>, this algorithm returns a supported configuration or <code>null</code> as appropriate.</p>
-            <p class="note">Unrecognized dictionary members in <var>candidate configuration</var> are ignored per [[WebIDL]] and will never reach this algorithm. Thus, they cannot be considered as part of the configuration.</p>
+            <p>Given a <a def-id="keysystems"></a> implementation <var>implementation</var>, <a>MediaKeySystemConfiguration</a> <var>candidate configuration</var>, and <var>origin</var>, this algorithm returns a supported configuration or <code>NotSupported</code> as appropriate.</p>
+            <p class="note">Unrecognized dictionary members in <var>candidate configuration</var> are ignored per [[WebIDL]] and will never reach this algorithm. Thus, they cannot be considered as part of the configuration.
+            </p>
+            <div class="note">
+              <p>
+              For certain configurations, it may be required to obtain user consent or inform the user. User Agents have some flexibility to determine
+              whether consent is required for a specific configuration and whether such consent may also apply to other configurations. For example,
+              consent to one configuration may also imply consent for less powerful, more restricted configurations. Equally, a denial of consent for
+              one configuration may imply denial of consent for more powerful, less restricted configurations.
+              </p>
+              <p>
+              Supported configurations, including supported audio and video codecs, may depend on availability of optional capabilities such as
+              <a def-id="distinctive-identifier"></a>s and persistent state. The following algorithm iteratively tries configurations with different
+              requirement values for these optional configuration elements until consent is achieved.
+              </p>
+              <p>
+              User Agents should persist consent responses at least for the duration of the <a def-id="requestMediaKeySystemAccess"></a>
+              algorithm in order to avoid repeated requests to the user for similar configurations.
+              </p>
+            </div>
+            <ol>
+              <li>
+                <p>Let <var>requirements</var> be the list [ <code>distinctiveIdentifier</code>, <code>persistentState</code> ].</p>
```

Comment:
I'm still thinking about this, but I'll post my current thoughts for discussion.

<br/>One thing I forgot to mention: It seems unlikely that a "not-allowed" value would require consent if a "required" value does not. As a bit of a corollary, I don't think denied consent should enable alternate options. (These are more gut instinct than definitive statements.)

Thus, do we need to run through all the combinations (or can we skip at least some based on denied consent)? (It's possible the observable behavior of the current text in the PR is the same, though it's hard to verify this currently.)


<br/>Thinking through possible values and implementations:

The "worst case" for the algorithm is that both are "optional". In that case, an implementation might choose [something that results in] "required" for DI and prompt. (PS could be either value.) If the consent is declined, that almost certainly means that DI should not be "not-allowed".

However, let's say the implementation chose [something that results in] "required" for PS too and asked for consent to both DI and PS (and maybe something else), or even just PS. In that case, it is _theoretically_ possible that the implementation could decide that PS should be "not-allowed" and it should retry with DI as "optional".

In the next loop, the implementation can choose [functionality that results in] any value for DI. In the worst case, it picks "required" again and prompts again. In another scenario, DI might only work with PS, resulting in the selection of "not-allowed". (I can't see how this could happen given the explicit separation of the two types of "persistence.") Perhaps the implementation still wants to prompt even though both DI and PS are "not-allowed".

All these combinations are possible if we iterate through all possible combinations. However, the order in which they can be tried is also prescribed, preventing the user agent from trying in its preferred order.


<br/>Perhaps most importantly, relying solely on varying the `MediaKeysRequirement` values also precludes varying other implementation-specific properties that could require consent.


<br/>Perhaps we can simplify this by allowing implementations to request consent only once per configuration. (As you said, how this is handled is up to the implementation.) Specifically:

1. Let the current algorithm run once with all the existing handling of "optional", order of importance, etc.
1. If consent is denied, let _restrictions_ = the set of things not consented to.
1. Update the "optional" requirements based on _restrictions_.
1. Rerun the algorithm exactly once with the new values.

This doesn't handle lack of consent for other things. To address that, we could pass a "_restrictions_ object" (empty in the first run) and check it along with _accumulated configuration_ in each such step (i.e. "...in combination with _accumulated configuration_ **and _restrictions_**").

While this _may_ be more limiting, it is clearer and more descriptive of how we intend for things to work (it's hard to verify that the current normative text in this PR has the same or similarly desired outcome) and allows consideration of implementation-specific properties. Also, the counterexamples for multiple rounds or multiple prompts seem unlikely, though we should acknowledge that we can't predict all implementations, especially those that might have multiple pipelines.